src/service/db/short_url.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright 2024 Zinc Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use config::utils::json;
+use infra::db::Event;
+
+use crate::{
+    common::infra::config::URL_MAP,
+    service::db::{self},
+};
+
+// DBKey to set short URL's
+pub const SHORT_URL_KEY: &str = "/short_urls/";
+
+pub async fn get(short_id: &str) -> Result<String, anyhow::Error> {
+    match URL_MAP.get(short_id) {
+        Some(val) => Ok(val.to_string()),
+        None => {
+            let val = db::get(&format!("{SHORT_URL_KEY}{short_id}"))
+                .await
+                .map_err(|_| anyhow!("Short URL not found in db"))?;
+
+            let original_url: String = json::from_slice(&val)
+                .map_err(|_| anyhow!("Failed to deserialize original url from db"))?;
+
+            URL_MAP.insert(short_id.to_string(), original_url.clone());
+
+            Ok(original_url)
+        }
+    }
+}
+
+pub async fn set(short_id: &str, original_url: &str) -> Result<(), anyhow::Error> {
+    URL_MAP.insert(short_id.to_string(), original_url.to_string());
+
+    db::put(
+        &format!("{SHORT_URL_KEY}{short_id}"),
+        json::to_vec(&original_url).unwrap().into(),
+        db::NEED_WATCH,
+        None,
+    )
+    .await?;
+
+    Ok(())
+}
+
+pub async fn get_by_original_url(original_url: &str) -> Option<String> {
+    if let Some(short_id) = URL_MAP.iter().find_map(|entry| {
+        let (k, v) = entry.pair();
+        if v == original_url {
+            return Some(k.clone());
+        }
+        None
+    }) {
+        return Some(short_id);
+    }
```

Comment:
if you have million record, this will be very slow, i am thinking we should use other solution, for example:
1. we get the hashing of original_url
2. get by short_url

Reply:
In our current operation, we have a SHORT_URLS in-memory map that stores <short_id, original_url>.
Would it makes sense to introduce another map that stores <original_url, short_id> for faster look up.

Need to the lookup for original_url since before shortening a url, need to check if one is already present.

Reply:
Have added `ORIGINAL_URLS` cache for faster reverse look up. Both `SHORT_URLS` & `ORIGINAL_URLS` are cleaned during gc.

Reply:
Modified the code such that it no longer needs to use `get_by_original_url`